[resol] Add Resol Controller Binding - Initial Contribution

ramack opened the pull request:
Target of this PR is to contribute the Resol binding to openhab and close https://github.com/openhab/openhab-addons/issues/7921 to support connection of devices from RESOL - Elektronische Regelungen GmbH, which are solar and heating controllers and heat count units. This includes support of many branded versions from Viessmann, SOLEX, COSMO, SOLTEX, DeDietrich and others.

The binding is already around and in production use since some years, discussed in the [community forum](https://community.openhab.org/t/beta-of-resol-binding-available-on-iot-marketplace/43550) and available in [Eclipse IoT Marketplace](https://marketplace.eclipse.org/content/resol-binding). Now it is the time to clean it up and submit a PR into openhab.

This is a reworked version for OH3 of what I submitted as https://github.com/openhab/openhab-addons/pull/8302 for the 2.5 branch.

ramack commented:
@fwolter @wborn @Hilbrand you have had comments on the 2.x variant of this PR, would you like to comment here, too?

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/ResolBindingConstants.java:
```diff
@@ -0,0 +1,55 @@
+
+    public static final ThingTypeUID THING_TYPE_UID_EMU_EM = new ThingTypeUID(BINDING_ID, THING_ID_EMU_EM);
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.unmodifiableSet(
+            Stream.of(THING_TYPE_UID_BRIDGE, THING_TYPE_UID_DEVICE, THING_TYPE_UID_EMU_EM).collect(Collectors.toSet()));
```
Same for below.
```suggestion
    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_UID_BRIDGE, THING_TYPE_UID_DEVICE, THING_TYPE_UID_EMU_EM);
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/ResolHandlerFactory.java:
```diff
@@ -0,0 +1,97 @@
+    private @Nullable LocaleProvider localeProvider;
+
+    @Reference
+    protected void setLocaleProvider(final LocaleProvider localeProvider) {
```
You could inject it into the constructor of this class to get rid of the Nullable annotation.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolDiscoveryService.java:
```diff
@@ -0,0 +1,90 @@
+            Map<String, Object> properties = new HashMap<>(1);
+            properties.put("type", type);
+
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withBridge(bridgeUID)
+                    .withProperties(properties).withLabel(name).build();
```
Can you add the representation property?

ramack replied:
For the thing "device" it there is only one of each type on each VBUS and therefore with the bridge unique. Can you please check whether I understood it right, now to do it then?

fwolter replied:
You've done it right!

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java:
```diff
@@ -0,0 +1,106 @@
+    /*
+     * The runnable for the search routine.
+     */
+    private Runnable searchRunnable = new Runnable() {
+
+        @Override
+        public void run() {
```
See above. Method.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java:
```diff
@@ -0,0 +1,106 @@
+        properties.put("adapterSerial", adapterSerial);
+
+        ThingUID uid = new ThingUID(ResolBindingConstants.THING_TYPE_UID_BRIDGE, adapterSerial);
+        thingDiscovered(
+                DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(dsWithInfo.getName()).build());
```
Can you add the representation property?

ramack replied:
The bridge is uniquely identified by the adapterSerial, which is read on discovery, but not required in file based configuration.
If I add a bridge manually in the configuration files the very same one is still autodetected, as long as I don't give the optional parameter adapterSerial. - Wouldn't it be better in this case to use the IP address as representation property?

fwolter replied:
Not sure. I would go for the IP address.

ramack replied:
I though I read that it should be a not-changing property and the IP would even not be recommended. But 1. I don't find that statement anymore, 2 the scenario that a bridge is using DHCP and getting a changed IP is quite unlikely.

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>Resol Binding</name>
+	<description>This is the binding for Resol solar and system controllers (including branded versions).</description>
+	<author>Raphael Mack</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/i18n/resol_xx_XX.properties:
```diff
@@ -0,0 +1,17 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
Can this file be removed?

ramack replied:
I was planning to do the translation into German at least and the file is a form of a reminder. But obviously it did not work well...

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,41 @@
+				<description>The IP address of the of the VBus-LAN adapter.</description>
+			</parameter>
+			<parameter name="port" type="integer" required="false" min="1024" max="65535">
+				<context>network-address</context>
```
This context is not valid for the port.

ramack replied:
is there a better one or can I simply remove the context tag completely?

fwolter replied:
Unfortunately there's no context for ports. So, you can remove it.

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,41 @@
+			</parameter>
+			<parameter name="port" type="integer" required="false" min="1024" max="65535">
+				<context>network-address</context>
+				<label>VBus-LAN adapter live data port</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,41 @@
+				<default>7053</default>
+			</parameter>
+			<parameter name="adapterSerial" type="text" required="false">
+				<label>AdapterSerial</label>
```
```suggestion
				<label>Adapter Serial Number</label>
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,41 @@
+			<parameter name="refreshInterval" type="integer" required="false" min="5" max="600" unit="s">
+				<context>refresh</context>
+				<label>Refresh Interval</label>
+				<description>Refreshtime in seconds for discovering devices, not relevant for data updates.</description>
```
```suggestion
				<description>Refresh time in seconds for discovering devices, not relevant for data updates.</description>
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,88 @@
+    <channel-type id="relay" advanced="false">
+        <item-type>Number:Dimensionless</item-type>
+        <label>Relay State</label>
+        <description>Virtual relay output, will be set by the controller and can be used to communicate data from the resol controller to Openhab.</description>
```
```suggestion
        <description>Virtual relay output, will be set by the controller and can be used to communicate data from the resol controller to openHAB.</description>
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,88 @@
+		<description>Solar or system controller (or any other real device on the VBus) from Resol.</description>
+	</thing-type>
+
+	<thing-type id="emulatedEM">
```
What do you mean by emulated?

ramack replied:
Normally an "EM" is a physical device which extends the solar controller by some sensor and relay channels and it is connected to the VBUS. This thing type allow openHAB to act as such a slave device on the VBUS and transmit sensor inputs and read relay output states to and from the solar controller. If a physical EM device is attached openHAB can only read the relay output and the sensor inputs measured by it but not transmit temperature values which will be handled by the solar controller.

ramack replied:
normally an EM device is a physical unit which extends the solar controller by 6 more sensor inputs and 5 relay outputs. If those are transmitting data on the VBUS they are discovered and openHAB is able to read the sensor inputs and relay output states.

The target of emulating an EM is to make openHAB a slave on the VBUS to be able to actively write sensor data which will then be used as "module sensors" in the solar controller, and the solar controller does not know that those values are injected by openHAB but will "see" them as another EM.

fwolter commented on bundles/pom.xml:
```diff
@@ -5,547 +5,1074 @@
   <modelVersion>4.0.0</modelVersion>
 
   <parent>
+
```
Can you remove the new lines?

fwolter replied:
The empty lines are still present.

fwolter reviewed (changes requested):
Thanks for your contribution again! I reviewed your code and here is my feedback.

There are some formatting issues. You can fix them with `mvn spotless:apply`.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

Can you update the year in the file headers?

What is the specification file with the .vsf extension about?

ramack commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,196 @@
+| Error_Sensor_line_broken-str  | String                   | Sensor line broken status (details for Error_mask) |
+| Flow_rate_sensor_x            | Number:VolumetricFlowRate| of sensor 'x'                                      |
+| Flow_set_temperature          | Number:Temperature       | Heating circuit set temperature                    |
+| Operating_state               | Number                   | Heating circuit operationg state                   |
+| Operating_state-str           | String                   | Heating circuit operationg state, as text          |
```
No I didn't. I searched for such a concept (looking for enum) but didn't find the state options. - This seems to do the trick, while it's not obvious how localization should be handled. - The string channel comes with the localized text (in en, de, anf fr) already from the library. - And it could be quite some effort to change but probably worth the effort. I'll investigate - sadly tomorrow I'll be back in work :-)

ramack commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,196 @@
+
+| channel                       | type                     | description                                        |
+|-------------------------------|--------------------------|----------------------------------------------------|
+| Pump_speed_relay_x            | Number:Dimensionless     | Percentage of the output state of relay 'x'        |
```
> Thanks for your contribution again! I reviewed your code and here is my feedback.

Thanks a lot. Some changes will be quite easy, others take some time but I'll tackle them in the upcoming days.

> What is the specification file with the .vsf extension about?
The used library vbus-java is using this internal file as the definition of communication on the resol VBUS. It is included in the library jar but for whatever reason the lib doesn't find it in it's own distribution when integrated into the openHAB build system. Maybe it automagically works now without since the library is included via mvn instead of copying the jar or you could give me some advice how to configure a resource search path to integrate the jar of the library to locate this file?

ramack replied:
ok. I use the name as defined by Resol and output by the vbus-java library and replace spaces and some other characters which are not allowed as channel names. Should I make them all lower case or just the first char?

fwolter replied:
If you use underscores to separate the words, all letters should be lower case.

ramack replied:
the underscores are replacements of spaces from the name in the library. But you're fully right it doesn't look good and I will just make all channel names lower case.

ramack commented:
> Thanks for your contribution again! I reviewed your code and here is my feedback.

Thanks a lot. Some changes will be quite easy, others take some time but I'll tackle them in the upcoming days.

> What is the specification file with the .vsf extension about?
The used library vbus-java is using this internal file as the definition of communication on the resol VBUS. It is included in the library jar but for whatever reason the lib doesn't find it in it's own distribution when integrated into the openHAB build system. Maybe it automagically works now without since the library is included via mvn instead of copying the jar or you could give me some advice how to configure a resource search path to integrate the jar of the library to locate this file?

fwolter commented on bundles/org.openhab.binding.resol/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+	Copyright (c) 2010-2020 Contributors to the openHAB project
+
+	See the NOTICE file(s) distributed with this work for additional
+	information.
+
+	This program and the accompanying materials are made available under the
+	terms of the Eclipse Public License 2.0 which is available at
+	http://www.eclipse.org/legal/epl-2.0
+
+	SPDX-License-Identifier: EPL-2.0
+
+-->
```
The feature.xml doesn't need a license header. You could remove it.

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,88 @@
+			<bridge-type-ref id="vbuslan"/>
+		</supported-bridge-type-refs>
+
+		<label>Emulated Extension Module (EM) Device</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,88 @@
+        
+        <config-description>
+            <parameter name="deviceId" type="integer" required="true" min="1" max="15">
+                <label>Module ID to be emulated</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,88 @@
+        <label>Temperature</label>
+        <description>Virtual temperature sensor, will be read by the controller and can be used to communicate data from Openhab to the resol controller.</description>
+        <category>Temperature</category>
+        <state pattern="%d %unit%" readOnly="true"/>
```
%d would result in conversion errors/warnings. Same for below.
```suggestion
        <state pattern="%.1f %unit%" readOnly="true"/>
```

fwolter commented:
> The used library vbus-java is using this internal file as the definition of communication on the resol VBUS. It is included in the library jar but for whatever reason the lib doesn't find it in it's own distribution when integrated into the openHAB build system. 

Yes, it would be good if this file could be read from the dependent JAR.

ramack commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,196 @@
+
+The channels of a thing are determined automatically based on the received VBUS data and are highly dependent on the used device.
+Here is a list of the channels of a DeltaSol MX with a heat quantity meter (HQM) and an extension module EM.
+The channels supported for your device can be seen after autodiscovery or 
```
good question :-) I have no clue how we should check it. Should we log the names in trace mode? - currently in trace the localized names - which are used as channel labels are printed. Maybe it would be better to change it output the channel name? Or should we just assume that any user is capable of looking the channels up in the web UI? - even those who still configure using text files?

I am tending to change the sentence to
**The channels supported for your device can be seen in the UI after data is received from the physical device.**

fwolter replied:
I think logging in trace is ok for users who doesn't use an UI.

ramack commented:
> > The used library vbus-java is using this internal file as the definition of communication on the resol VBUS. It is included in the library jar but for whatever reason the lib doesn't find it in it's own distribution when integrated into the openHAB build system.
>
> Yes, it would be good if this file could be read from the dependent JAR.

seems to work now, probably since the mvn dependency is used

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,522 @@
+                            tcpConnection = null;
+                        }
+                    } catch (IOException e) {
+                        logger.warn("TCP disconnect failed", e);
```
no, not really. e.getMessage() should be sufficient.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,522 @@
+
+                        if (connection != null) {
+                            try {
+                                Thread.sleep(3000); // Wait for connection...
```
you are right. I guess this was left over from an early phase where the listener did not correctly update the connected status. Will remove it.

Hilbrand commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,522 @@
+    private Specification spec;
+    private Set<String> availableDevices = new HashSet<String>();
+
+    private Map<String, ResolThingHandler> thingHandlerMap = new HashMap<String, ResolThingHandler>();
```
You probably should not keep a map of things. The bridge already tracks this itself via children. So you're re-implementing part of what is already available, making this code more complex than needed. See if you can change the code to make use of this functionality.

ramack replied:
I feel stupid or blind right now :grinning:, can you please point me to it?

Hilbrand replied:
You can get it via `getThing().getThings()` see for example https://github.com/openhab/openhab-addons/blob/b6eea715fa74e630283726d603f5e24f9f315a6f/bundles/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyBridgeHandler.java#L427

Hilbrand commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolDiscoveryService.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    protected void startScan() {
+        // Scan will be done by bridge
```
This is not allowed. This service should be in control of when scanning is done. What you can do is instead of in activate  register to the bridge here and in stopScan unregister. Or keep a list of unknown things in the bridge and let discovery get (and clear) the list when discovery is started here.

Hilbrand commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/ResolHandlerFactory.java:
```diff
@@ -0,0 +1,97 @@
+        return null;
+    }
+
+    private synchronized void registerThingDiscovery(ResolBridgeHandler bridgeHandler) {
+        ResolDiscoveryService discoveryService = new ResolDiscoveryService(bridgeHandler);
+        logger.trace("Try to register VBUS Discovery service on BundleID: {} Service: {}",
+                bundleContext.getBundle().getBundleId(), DiscoveryService.class.getName());
+
+        Hashtable<@Nullable String, String> prop = new Hashtable<@Nullable String, String>();
+
+        bundleContext.registerService(DiscoveryService.class.getName(), discoveryService, prop);
+        discoveryService.activate();
+    }
```
Can you implement registering the discovery bound to a bridge as a `ThingHandlerService`. See this preliminary documentation on how to do that: https://deploy-preview-1262--openhab-docs-preview.netlify.app/docs/developer/bindings/#discovery-that-is-bound-to-a-thing

ramack replied:
ok, I tried that, can you please check it? What I recognized, was that bridge defined in a thing file was still detected (even though I changed the representationproperty to the ipAddress) but when I added the discovered one it magically disappeared... So maybe I didn't do it right fully.

Hilbrand reviewed (changes requested):
I've not looked at everything, but there are some conceptual things I wanted to address first.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,522 @@
+    }
+
+    /* adapter to react on connection state changes and handle received packets */
+    private class ResolConnectorAdapter extends ConnectionAdapter {
```
This will get quite a big refactoring. Do you think it is worth it?

fwolter replied:
It's only three clicks in Eclipse. It was only a suggestion. I won't insist on changing it.

ramack replied:
I tried that, but it failed. And the inner class accesses quite some outer class fields - maybe that is also not the nicest design but I would not want to change that without a strong need :grinning:

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,245 @@
+        }
+    }
+
+    @Override
+    public void updateStatus(ThingStatus status) {
+        super.updateStatus(status);
+    }
```
OMG - what did I do here? For sure this shall vanish.

ramack replied:
ah, now as I removed it the compiler reminds me why it is there... actually it changes the access from protected to public to allow the call from ResolBridgeHandler.registerResolThingListener()

ramack commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,196 @@
+| System_date                   | DateTime                 | Date and time of the controller clock              |
+| Error_mask                    | Number                   | Bitmask for the different errors                   |
+| Error_Sensor_line_broken-str  | String                   | Sensor line broken status (details for Error_mask) |
+| Flow_rate_sensor_x            | Number:VolumetricFlowRate| of sensor 'x'                                      |
```
But if I check here: https://next.openhab.org/docs/concepts/units-of-measurement.html
the type is given as VolumetricFlowRate - ok, I could still add "Flow rate " to the description, but the unit should stay, shouldn't it?

ramack replied:
@Hilbrand I don't catch the point here. Can you explain why Number:VolumetricFlowRate is not correct?

Hilbrand replied:
Yes you are right. I misread the description. Maybe indeed a good idea to add it.

ramack commented:
> I've not looked at everything, but there are some conceptual things I wanted to address first.

Thanks a lot for your review, will tackle it in the next few days, as it is bit more refactoring.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,169 @@
+        unregisterResolThingListener(bridgeHandler);
+    }
+
+    @Override
+    public ThingBuilder editThing() {
+        return super.editThing();
+    }
+
+    @Override
+    public void updateStatus(ThingStatus status) {
+        super.updateStatus(status);
+    }
+
+    @Override
+    public void updateThing(Thing thing) {
+        super.updateThing(thing);
+    }
```
same here to extend the access

ramack commented:
Rework after the first review is not yet finished, I just wanted to push the changes to github.

ramack commented:
@fwolter in https://github.com/openhab/openhab-addons/pull/9449/commits/781ed3f90af9492e6ef0a4a30b1a97de2aaba9ad I tried to as state options also for the static channel types for the emulatedEM but I don't manage to make the Main UI showing a drop down for it. Can you help me why?

Beside this the only thing left is a general rework to
* move the reponsibility of the package parsing into the Thing handlers
* remove the methods only calling super like (updateStatus adn editThing) in the thing handlers
* discovery mechanism to discover only during scan

fwolter commented:
If you want to send commands to the Channel, you need to use [command descriptions](https://www.openhab.org/docs/developer/bindings/thing-xml.html#command-description) instead of state options.

ramack commented:
> If you want to send commands to the Channel, you need to use [command descriptions](https://www.openhab.org/docs/developer/bindings/thing-xml.html#command-description) instead of state options.

This I don't catch. Setting the mode in channel looks for me like the logical correct thing, as it is more like a state than a single event. With the state options in https://github.com/ramack/openhab-addons/commit/781ed3f90af9492e6ef0a4a30b1a97de2aaba9ad this seems correct to me, but in the main UI I don't find how to configure a widget to show a drop down to select the possible options.

fwolter commented:
If you use command options, you can configure a Label Cell and set the action to "Command options":
![grafik](https://user-images.githubusercontent.com/1894841/105197488-f4e8e080-5b3c-11eb-8d34-8fd96b99f375.png)



If you click on the label it looks like this:
![grafik](https://user-images.githubusercontent.com/1894841/105197555-03cf9300-5b3d-11eb-9cc8-56c4c2491cfb.png)

Is that what you are looking for?

ramack commented:
> If you use command options, you can configure a Label Cell and set the action to "Command options":
> Is that what you are looking for?

Yes, exactly. Thanks for training me here for OH3. I didn't expect that if an item is configured as SetPoint I have to set the item two times in the configuration of the MainUI to write to the value, but at least the binding seems ok about how it uses the state options.

ramack commented:
I want to do some further testing and a rebase, but @Hilbrand and @fwolter can you already take a look whether I correctly considered your findings in the review?

ramack commented:
oh my goodness, what did I do? I only wanted to rebase the PR on top of openhab-main to solve the conflict in the pom, sorry for the mess. Can anybody help me to clean this up? - I don't understand what happend...

Hilbrand commented:
Either you forgot to add `--rebase` when you did git pull or you forgot to use `--force` when pushing to GitHub.

ramack commented:
> Either you forgot to add `--rebase` when you did git pull or you forgot to use `--force` when pushing to GitHub.

Thanks for the feedback. Before I do more harm, do you have  a trick for me how to fix it?

fwolter commented:
You used merge here 2be8eb3, which shouldn't be used in pull-requests.

You can reset your Git history before the merge: `git reset --soft 14e88d0`

Then, discard all changes from the merge and commit your intended changes again.

When everything looks fine, force push it: `git push --force`

fwolter commented:
Here are the commands for rebasing your branch onto `main`:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
The remotes (`git remote --verbose`) should look like this, now:
```
origin  https://github.com/[your name]/openhab-addons.git (fetch)
origin  https://github.com/[your name]/openhab-addons.git (push)
upstream        https://github.com/openhab/openhab-addons.git (fetch)
upstream        https://github.com/openhab/openhab-addons.git (push)
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```
**DON'T** use merge! This can clutter the Git history and make the PR unusable.

ramack commented:
The issue was that I did a pull instead of a rebase, was probably time for bed. Looks better now, but still we have the request for review pending for too many developers - can I do here something?

fwolter commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,250 @@
+
+A more complete list can be found in the doc of the [resol-vbus-java library](http://danielwippermann.github.io/resol-vbus/vbus-packets.html).
+
+###Emulated Extension Module EM
```
```suggestion
### Emulated Extension Module EM
```

fwolter commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,250 @@
+## Bridge Configuration
+
+The bridge is the device connecting the Resol VBUS to the network, usually a VBus-LAN adapter or integrated in some of the solar controllers like DL3.
+For the connection from the Resol binding the bridge requires the configuration of
```
```suggestion
For the connection from the Resol binding the bridge requires the configuration of the following parameters:
```

fwolter commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,250 @@
+The bridge is the device connecting the Resol VBUS to the network, usually a VBus-LAN adapter or integrated in some of the solar controllers like DL3.
+For the connection from the Resol binding the bridge requires the configuration of
+
+| property             | type    | Required | description                                                |
```
```suggestion
| parameter             | type    | Required | description                                                |
```

fwolter commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,250 @@
+
+## Emulated EM Configuration
+
+*emulatedEM* devices cannot be auto-discovered and require beside the bridge the following configuration:
```
```suggestion
*emulatedEM* devices cannot be auto-discovered and require beside the bridge the following configuration:

```

ramack replied:
What is different here? I am too blind to see the change from "and" to "and" :cry:

fwolter replied:
It's the new line after the colon. There should be an empty line before tables, head lines and so on.

ramack replied:
ah yes, clear. I was focused on the red indicated and in the diff such that I didn't recognize the newline, sorry.

fwolter commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,250 @@
+|-----------------------------------|--------------------------|----------------------------------------------------|
+| pump_speed_relay_x                | Number:Dimensionless     | Percentage of the output state of relay 'x'        |
+| temperature_sensor_x              | Number:Temperature       | Temperature sensor 'x' of the controller           |
+| temperature_Module_y_Sensor_x     | Number:Temperature       | Temperature sensor 'x' of the extension module 'y' |
```
```suggestion
| temperature_module_y_Sensor_x     | Number:Temperature       | Temperature sensor 'x' of the extension module 'y' |
```

fwolter commented on bundles/org.openhab.binding.resol/README.md:
```diff
@@ -0,0 +1,250 @@
+| output_m                          | Number:Dimensionless     | PWM/0-10V level value of the output 'm'            |
+| system_date                       | DateTime                 | Date and time of the controller clock              |
+| error_mask                        | Number                   | Bitmask for the different errors                   |
+| error_sensor_line_broken          | Number                   | Sensor line broken status (details for Error_mask) |
+| error_sensor_line_short-circuited | Number                   | Sensor short circuit status (details for Error_mask) |
```
```suggestion
| error_sensor_line_broken          | Number                   | Sensor line broken status (details for error_mask) |
| error_sensor_line_short-circuited | Number                   | Sensor short circuit status (details for error_mask) |
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,343 @@
+                            || ConnectionState.INTERRUPTED.equals(connState)) {
+                        isConnected = false;
+                    }
+                    logger.info("Connection state changed to: {}", connState.toString());
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,343 @@
+             * 0x10, because the MX redelivers the data from the EM to the DFA.
+             * But the MX is the exception in this case and many other controllers do not redeliver data, so we keep it.
+             */
+            logger.trace("Received Data from {} (0x{}/0x{}) naming it {}",
```
You could enclose this with an if, that the data is not calculated when running in normal logging mode.
```suggestion
            if (logger.isTraceEnabled()) {
                logger.trace("Received Data from {} (0x{}/0x{}) naming it {}",
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,297 @@
+    private long lastTime = System.currentTimeMillis();
+
+    // Background Runnable
+    @Nullable
+    private ScheduledFuture<?> updateJob;
```
```suggestion
    private @Nullable ScheduledFuture<?> updateJob;
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,297 @@
+        int intValue = 0;
+
+        if (command instanceof QuantityType<?>) {
+            value = ((QuantityType<?>) command).floatValue();
```
This will convert to the current quantity's unit. E.g. if the user uses Fahrenheit, it will convert to the Fahrenheit value. You can use `command.toUnit()` to convert it to the expected value.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,297 @@
+                if (ste.equals(ConnectionState.CONNECTED)) {
+                    updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
Can you set the status detail message, that the user gets a clue what is wrong?

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+    ResolBridgeHandler bridgeHandler;
+    private ResolStateDescriptionOptionProvider stateDescriptionProvider;
+
+    private SimpleDateFormat dateFormat = new SimpleDateFormat(DateTimeType.DATE_PATTERN_WITH_TZ_AND_MS_GENERAL);
```
Can this be a constant?

ramack replied:
Do you mean to make it final?

fwolter replied:
Yes, and make it `static`. `SimpleDateFormat` is also not thread-safe. You need to synchronize access if it can be accessed by multiple threads concurrently.

ramack replied:
ok. I don't see any reason here to have multithreaded calls here, so I'd not add synchronized.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,292 @@
+
+            if (pfv.getEnumVariant() != null) {
+                /* update the enum / State channel */
+                setChannelValue(channelId, pfv.getRawValueLong()); // pfv.getEnumVariant().getText(lang));
```
Can the comment be removed?

ramack replied:
it can not only, it shall be removed :grinning:

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,122 @@
+        super(Set.of(ResolBindingConstants.THING_TYPE_UID_DEVICE), 15, false);
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(ResolDeviceDiscoveryService.class);
+
+    private @Nullable ResolBridgeHandler resolBridgeHandler;
```
Fields should be placed at the top of the class.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    protected void startScan() {
+        scheduler.execute(this::searchRunnable);
```
This should be cancelled if `stopScan()` is invoked.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java:
```diff
@@ -0,0 +1,98 @@
+        properties.put("adapterSerial", adapterSerial);
+
+        ThingUID uid = new ThingUID(ResolBindingConstants.THING_TYPE_UID_BRIDGE, adapterSerial);
+        thingDiscovered(DiscoveryResultBuilder.create(uid).withRepresentationProperty("ipAddress")
```
Better make `ipAddress` a constant.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,122 @@
+            properties.put("type", type);
+
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withBridge(bridgeUID)
+                    .withRepresentationProperty("type").withProperties(properties).withLabel(name).build();
```
Better make `type` a constant.

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/i18n/resol_de.properties:
```diff
@@ -0,0 +1,45 @@
+# binding
+binding.resol.name = Resol Binding
+binding.resol.description = Verbinded Solar- und Systemregler des Herstellers Resol und vielen die für andere Marken von Resol produziert werden.
```
```suggestion
binding.resol.description = Verbindet Solar- und Systemregler des Herstellers Resol und weitere, die für andere Marken von Resol produziert werden.
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/i18n/resol_de.properties:
```diff
@@ -0,0 +1,45 @@
+thing-type.config.resol.vbuslan.adapterSerial.description = Seriennummer des VBus-LAN-Schnittstellengerätes (nur zur Information).
+thing-type.config.resol.vbuslan.password.label = Passwort
+thing-type.config.resol.vbuslan.password.description = Passwort für die VBus-LAN Schnittstelle.
+thing-type.config.resol.vbuslan.refreshInterval.label = Aktualisierungsinterval
```
```suggestion
thing-type.config.resol.vbuslan.refreshInterval.label = Aktualisierungsintervall
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/i18n/resol_de.properties:
```diff
@@ -0,0 +1,45 @@
+thing-type.config.resol.vbuslan.password.label = Passwort
+thing-type.config.resol.vbuslan.password.description = Passwort für die VBus-LAN Schnittstelle.
+thing-type.config.resol.vbuslan.refreshInterval.label = Aktualisierungsinterval
+thing-type.config.resol.vbuslan.refreshInterval.description = Zeitinterval in Sekunden um die Verbindung zur VBus-LAN-Schnittstelle zu prüfen. Die Aktualisierung der Daten geschieht unabhängig hiervon sobald sie auf dem VBus empfangen werden.
```
```suggestion
thing-type.config.resol.vbuslan.refreshInterval.description = Zeitintervall in Sekunden um die Verbindung zur VBus-LAN-Schnittstelle zu prüfen. Die Aktualisierung der Daten geschieht unabhängig hiervon sobald sie auf dem VBus empfangen werden.
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,42 @@
+				<context>refresh</context>
+				<label>Refresh Interval</label>
+				<description>Refresh time in seconds to check the connection to the VBus gateway. Data updates are propagated to
+					openHAB independent from this setting as soon as the are received on the VBus.</description>
```
```suggestion
					openHAB independently from this setting as soon as they are received on the VBus.</description>
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+        int intValue = 0;
+
+        if (command instanceof QuantityType<?>) {
+            value = Objects.requireNonNullElse(((QuantityType<?>) command).toUnit(SIUnits.CELSIUS),
+                    new QuantityType<>(888.8, SIUnits.CELSIUS)).floatValue();
```
What is the meaning of the value of 888.8°C? You could log a warning if the unit cannot be converted. This occurs if the user sent an invalid command.

ramack replied:
888.8 is used in the VBUS in case of not available values - e. g. if the sensor wire would be broken. So in my eyes it fits quite well to the illegal input case here.

fwolter replied:
Me as a user would be happy about a hint what went wrong. I leave it up to you how to handle that case.

fwolter commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,293 @@
+    ResolBridgeHandler bridgeHandler;
+    private ResolStateDescriptionOptionProvider stateDescriptionProvider;
+
+    private static final SimpleDateFormat dateFormat = new SimpleDateFormat(
```
As it is a constant now, the name should be all upper case.

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/i18n/resol_de.properties:
```diff
@@ -0,0 +1,45 @@
+# binding
+binding.resol.name = Resol Binding
+binding.resol.description = Verbindet Solar- und Systemregler des Herstellers Resol und weiteren, die für andere Marken von Resol produziert werden.
```
```suggestion
binding.resol.description = Verbindet Solar- und Systemregler des Herstellers Resol und weitere, die für andere Marken von Resol produziert werden.
```

fwolter commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,42 @@
+				<context>refresh</context>
+				<label>Refresh Interval</label>
+				<description>Refresh time in seconds to check the connection to the VBus gateway. Data updates are propagated to
+					openHAB independently from this setting as soon as the are received on the VBus.</description>
```
```suggestion
					openHAB independently from this setting as soon as they are received on the VBus.</description>
```

fwolter reviewed (approved):
LGTM

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

ramack commented:
> checkstyle

ramack commented:
@Hilbrand do you see any chance to take a second look?

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,124 @@
+ *
+ * @author Raphael Mack - Initial contribution
+ */
+@Component(service = DiscoveryService.class)
+@NonNullByDefault
+public class ResolDeviceDiscoveryService extends AbstractDiscoveryService
+        implements DiscoveryService, ThingHandlerService {
```
The `@Component` annotation isn't compatible with `ThingHandlerService` atm.
```suggestion
@NonNullByDefault
public class ResolDeviceDiscoveryService extends AbstractDiscoveryService
        implements DiscoveryService, ThingHandlerService {
```

ramack replied:
> The code looks pretty good to be honest. Only a few changes I'd suggest.

Thanks, that is due to the valuable review comments from @fwolter and @Hilbrand!

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,297 @@
+
+    private final Logger logger = LoggerFactory.getLogger(ResolThingHandler.class);
+
+    @Nullable
+    ResolBridgeHandler bridgeHandler;
```
Is there a reason this needs to be package private?

ramack replied:
no, and we even can get rid of this completely.

cpmeister reviewed (changes requested):
The code looks pretty good to be honest. Only a few changes I'd suggest.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,297 @@
+
+    public ResolThingHandler(Thing thing, ResolStateDescriptionOptionProvider stateDescriptionProvider) {
+        super(thing);
+        synchronized (DATE_FORMAT) {
+            DATE_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
+        }
```
makes sense, yes. But do I need to keep the synchronized block in the static initializer? Without having that deep experience in Java I would expect, that the static init is ensured to be executed completely before any other thread accesses methods of objects of this class. - Is that true, or should I keep the synchronized even there?

ramack replied:
@cpmeister any more comment on this one? Keep the synchronized in the static initialization or is it not needed there?

ramack commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,142 @@
+			<channel id="relay_2" typeId="relay"/>
+			<channel id="relay_3" typeId="relay"/>
+			<channel id="relay_4" typeId="relay"/>
+			<channel id="relay_5" typeId="relay"/>
```
yes per EM there are 6 sensors inputs and 5 relays on EM devices.

Hilbrand commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+                    if (intValue == 0) {
+                        /* switch is open => 1 megaohm */
+                        dev.setResistorValueByNr(channel, 1000000000);
+                        updateState(channelUID, OnOffType.OFF);
```
Why update the channels here? at this point it gets a command because the value of the channel changed. So this update changes the value to the value the channel already contains? So it seems redundant to do this.

ramack replied:
I added these updates, because otherwise the MainUI did not correctly update the view (of the widget I tried with). Maybe I just did something wrong as I am not yet used to MainUI or whatever. Let me know if I should remove it.

ramack replied:
@Hilbrand any remark on that? - I have added the update here to have the MainUI showing the update. Should I remove it?

Hilbrand commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/ResolHandlerFactory.java:
```diff
@@ -0,0 +1,78 @@
+@Component(configurationPid = "binding.resol", service = ThingHandlerFactory.class)
+public class ResolHandlerFactory extends BaseThingHandlerFactory {
+
+    private LocaleProvider localeProvider;
```
These fields can be final

Hilbrand commented on bundles/org.openhab.binding.resol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,142 @@
+		<description>Emulation of an Extension Module (EM) device which can be connected through the VBUS to Resol controllers
+			which support the EM devices. Replaces a physically available EM by openHAB.</description>
+		<channels>
+			<channel id="relay_1" typeId="relay"/>
```
You can give the enumerated channels a label so they are better recognizable in the ui:
```suggestion
			<channel id="relay_1" typeId="relay">
			    <label>Relay 1</label>
			</channel>
```

ramack replied:
ok, should in this case the Label in the channel-type be removed?

ramack replied:
I have added the individual labels, but should the label in the channel-type stay as it is never used now?

Hilbrand reviewed (changes requested):
Some last review comments.

ramack commented:
@Hilbrand and @cpmeister do you have any further proposals for improvement on which I could work?

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
```
Can you also handle refresh commands?

ramack replied:
I don't think so, as there is no polling mechanism on the VBUS, we have to wait until the packets are transmitted the next time.

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/ResolHandlerFactory.java:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        if (thingTypeUID.equals(ResolBindingConstants.THING_TYPE_UID_BRIDGE)) {
+            ResolBridgeHandler handler = new ResolBridgeHandler((Bridge) thing, localeProvider);
+            return handler;
```
```suggestion
            return new ResolBridgeHandler((Bridge) thing, localeProvider);
```

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java:
```diff
@@ -0,0 +1,113 @@
+    @Override
+    protected void startScan() {
+        discoveryRunning = true;
+        scheduler.execute(this::searchRunnable);
```
Can you point me to an example? Since I was requested to use directly the execute() with the method as parameter I don't have any object anymore that could be canceled...

cpmeister replied:
If you use `submit` instead then it would return a Future which you could store to a field and then cancel that future as part of `stopScan`.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+            value = intValue;
+            update = true;
+        } else {
+            update = false;
```
By a return you mean? Is that really better or purely a matter of taste?

cpmeister replied:
Well by returning early you don't need to use an `update` variable. I'd say that less variables would make the code more maintainable.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    @Override
+    public void dispose() {
+        EmDeviceEmulator dev = device;
+        if (dev != null) {
+            dev.stop();
+            dev.removePropertyChangeListener(this);
+        }
+    }
```
how do I do that?

cpmeister replied:
You already do the canceling as part of `ResolEmuEMThingHandler.stop()` so you either need to call that method or make a common method that does the canceling logic and call that instead.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolBridgeHandler.java:
```diff
@@ -0,0 +1,345 @@
+        super(bridge);
+        spec = Specification.getDefaultSpecification();
+
+        if (localeProvider != null) {
+            locale = localeProvider.getLocale();
+            lang = SpecificationFile.getLanguageForLocale(getLocale());
+
+        } else {
+            locale = Locale.getDefault();
+            lang = Language.En;
+        }
```
Is it worth to increase the depth of the call tree for each received packet for checking the locale only for this rather seldom use case? As a user, I personally would favor runtime performance over this. Or is there at least a callback that I can register to get notified on the locale update?

cpmeister replied:
The performance impact would be negligible and would allow the user to change the locale in openhab without having to restart your binding.
That said I don't know what the performance impact of calling `SpecificationFile.getLanguageForLocale` would be. So it might be worth caching the return value from that method and only changing it when the returned locale from the provider changes.

ramack replied:
`SpecificationFile.getLanguageForLocale` looks also quite fast, so I changed to read the locale dynamically.

ramack commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java:
```diff
@@ -0,0 +1,113 @@
+                    currentDataSourceById.put(addressId, dsWithInfo);
+                    addAdapter(addressId, dsWithInfo);
+                    // here we can add the detection of Multi-Channel interfaces like DL3
+                } catch (IOException ex) {
+                    /* address is no valid adapter */
+                }
```
and why should we even? If for one of the TcpDataSource's that we get from the discovery in the library one is faulty, why should we not try the next one?

cpmeister replied:
I meant that you should add an extra catch statement for `InterruptedIOException` which is a type of `IOException` which is thrown when the thread doing IO is interrupted. The reason for requiring this is to make sure that the thread actually exits as fast as possible when an interrupt occurs. When a thread receives and interrupts it should be assumed that the system is trying to shutdown and typically complete shutdown can't occur until all of the non-daemon threads have stopped running. Since the running thread might be non-daemon it prevent the system from shutting down as long as it is running.

ramack replied:
ok, this way I think I got it! Thanks for the explanation.

ramack commented:
Thanks @cpmeister for the the review, I reworked some findings already, for others I have added some questions and I'm looking forward to your response!

ramack commented:
The jenkins build failed, but I don't catch why - there is too much in there that I do not understand. Can I do something to fix it?

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,298 @@
+        if (dev != null) {
+            if (chID.startsWith(CHANNEL_TEMP)) {
+                dev.setResistorValueByNrAndPt1000Temperatur(channel, value);
+                updateState(channelUID, new DecimalType(value));
```
You should be using QuantityType here.

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolEmuEMThingHandler.java:
```diff
@@ -0,0 +1,298 @@
+                    dev.setResistorValueByNr(channel, 0);
+                    updateState(channelUID, OnOffType.ON);
+                }
+            } else if (chID.startsWith(CHANNEL_RESIST)) {
+                dev.setResistorValueByNr(channel, (int) (value * 1000.0));
+                updateState(channelUID, new QuantityType<>(intValue, Units.OHM));
+            } else if (chID.startsWith(CHANNEL_TEMP_ADJUST)) {
+                basValues[channel - 1].temperatureOffset = value;
+                updateBas(channel);
+                updateState(channelUID, new DecimalType(value));
```
and here

cpmeister commented on bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/handler/ResolThingHandler.java:
```diff
@@ -0,0 +1,297 @@
+    protected void packetReceived(Specification spec, Language lang, Packet packet) {
+        PacketFieldValue[] pfvs = spec.getPacketFieldValuesForHeaders(new Packet[] { packet });
+        for (PacketFieldValue pfv : pfvs) {
+            logger.trace("Id: {}, Name: {}, Raw: {}, Text: {}", pfv.getPacketFieldId(), pfv.getName(lang),
+                    pfv.getRawValueDouble(), pfv.formatTextValue(null, Locale.getDefault()));
```
Actually at second look. It seems that I was mistaken on this one.

ramack commented:
Sorry for the long time until this update.

ramack commented:
@cpmeister @Hilbrand is there anything left I should do for this PR?

fwolter commented:
Due to lack of review capacity, the rules for new bindings were weakened. New bindings need only the approval of one reviewer instead of two from now on.

EjvindHald commented:
@ramack Really great that this binding is going to work also in OH cer 3.x

Have you verified that it can be configured only by using a thing + item file? So not using any setup in OH ui.

Have you verified that the values from Resol are refreshed also when the binding is used from addons folder?

Both of these 2 above were a bit problematic in Resol binding ver 2.x.

Thanks a lot for your effort.

ramack commented:
Yes, I am using the binding in my productive setup with only file based configuration. Your second point about the addons folder I don't fully catch - do you refer to installing the jar manually compared to getting it from the marketplace? - In this case we are also safe.
But nothing to thanks me, the main work was done by @danielwippermann in the used library.

Looking forward to get further feedback from the next milestone build.

EjvindHald commented:
> do you refer to installing the jar manually compared to getting it from the marketplace?

Yes, in version 2.x only installs from market place worked fine with no problems. However, If the jar file was placed in the addons folder, it did not work properly. Readings from Resol devices were not updated - only set upon start of OH 2.x.

ramack commented:
> > do you refer to installing the jar manually compared to getting it from the marketplace?
> 
> Yes, in version 2.x only installs from market place worked fine with no problems. However, If the jar file was placed in the addons folder, it did not work properly. Readings from Resol devices were not updated - only set upon start of OH 2.x.

ok, so no worries. Marketplace is dead. The binding in OH3 works ONLY if the jar is in the addons folder :-)
